src/dbnode/storage/series/buffer.go

```diff
@@ -473,119 +460,326 @@ func (b *dbBuffer) FetchBlocksMetadata(
+			merges, err := bucket.merge()
+			if err != nil {
+				return res, nil
+			}
+			res += merges
+		}
+	}
+
+	return res, nil
 }
 
-func (b *dbBufferBucket) empty() bool {
-	for _, block := range b.bootstrapped {
-		if block.Len() > 0 {
-			return false
+func (b *BufferBucketVersions) removeBucketsUpToVersion(version int) {
+	// Avoid allocating a new backing array.
+	nonEvictedBuckets := b.buckets[:0]
+
+	for _, bucket := range b.buckets {
+		bVersion := bucket.version
+		// TODO(juchan): deal with ColdWrite too.
+		if bucket.wType == WarmWrite && bVersion != writableBucketVer &&
+			bVersion <= version {
+			// We no longer need to keep any version which is equal to
+			// or less than the retrievable version, since that means
+			// that the version has successfully persisted to disk.
+			// Bucket gets reset before use.
+			b.bucketPool.Put(bucket)
+			continue
 		}
+
+		nonEvictedBuckets = append(nonEvictedBuckets, bucket)
 	}
-	for _, elem := range b.encoders {
-		if elem.encoder != nil && elem.encoder.NumEncoded() > 0 {
-			return false
+
+	b.buckets = nonEvictedBuckets
+}
+
+func (b *BufferBucketVersions) setLastRead(value time.Time) {
+	atomic.StoreInt64(&b.lastReadUnixNanos, value.UnixNano())
+}
+
+func (b *BufferBucketVersions) lastRead() time.Time {
+	return time.Unix(0, atomic.LoadInt64(&b.lastReadUnixNanos))
+}
+
+func (b *BufferBucketVersions) bootstrap(bl block.DatabaseBlock) {
+	// TODO(juchan): what is a "cold" bootstrap?
+	bucket := b.writableBucketCreate(WarmWrite)
+	bucket.bootstrapped = append(bucket.bootstrapped, bl)
+}
+
+func (b *BufferBucketVersions) writableBucket(wType WriteType) (*BufferBucket, bool) {
+	for _, bucket := range b.buckets {
+		if bucket.version == writableBucketVer && bucket.wType == wType {
+			return bucket, true
 		}
 	}
-	return true
+
+	return nil, false
 }
 
-func (b *dbBufferBucket) canRead() bool {
-	return !b.drained && !b.empty()
+func (b *BufferBucketVersions) writableBucketCreate(wType WriteType) *BufferBucket {
+	bucket, exists := b.writableBucket(wType)
+
+	if exists {
+		return bucket
+	}
+
+	newBucket := b.bucketPool.Get()
+	newBucket.resetTo(b.start, wType, b.opts)
+	b.buckets = append(b.buckets, newBucket)
+	return newBucket
 }
 
-func (b *dbBufferBucket) needsReset(
-	targetStart time.Time,
-) bool {
-	return !b.start.Equal(targetStart)
+func (b *BufferBucketVersions) toStreams(ctx context.Context) ([]xio.SegmentReader, error) {
+	buckets := b.buckets
+	res := make([]xio.SegmentReader, 0, len(buckets))
+
+	for _, bucket := range buckets {
+		stream, err := bucket.toStream(ctx)
+		if err != nil {
+			return nil, err
+		}
+		res = append(res, stream)
+	}
+
+	return res, nil
 }
 
-func (b *dbBufferBucket) needsDrain(
-	now time.Time,
-	targetStart time.Time,
-) bool {
-	retentionOpts := b.opts.RetentionOptions()
-	blockSize := retentionOpts.BlockSize()
-	bufferPast := retentionOpts.BufferPast()
+// BufferBucket is a bucket in the buffer.
+type BufferBucket struct {
```

Comment:
Does this need to be exported?

Reply:
Since we want to have a database-level pool of these, we use this when passing pools via our Options.